design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,433 @@
+<!-- is this true? what should Ceph do before replacement, mark osd out? -->. Rook will remove the
+OSD as soon as Ceph reports the OSD safe to remove. While the user is waiting on the OSD to be
+removed by Rook, they will go about removing the old disk(s) from the node and inserting a new
+disk(s). When the old OSDs Kubernetes resources have been deleted and the new disk(s) is inserted,
+Rook will recreate the OSD on the new disk(s).
+
+The queue item will have to contain the information necessary for Rook to delete the old disk's
+Kubernetes resources, the information necessary to determine when a new disk is present that is
+suitable for replacing the old disk, and the information necessary to create a new OSD with the same
+OSD ID as the one being replaced.
+
+The Ceph mgr may crash after the replacement has been queued but before Ceph marks the OSD out.
+Because of this, the mgr should inspect all replacement operations in the queue at startup and mark
+OSDs out if necessary.
+
+#### Listing OSDs and available devices on hosts
+The Ceph mgr will be able to view the **stage 1** inspection report ConfigMap to get a list of
+devices on each host. The user or Ceph orchestrator may wish to receive an updated list of devices.
+Using the "orchestrator-osd-requests" ConfigMap, Ceph will indicate that it wishes Rook to update
+its device list. When the control loop is not creating or destroying an OSD, it should kick off
+**stage 1** to list devices on all nodes and then indicate that it has completed the device list
+update, at which time the Ceph orchestrator can now read the most up-to-date inspection report valid
+for all nodes. Updating the device list takes priority over creation, deletion, and replacement.
+
+> **TODO**
+> how will we handle atomicity of configmap updates?
+>
+> one option would be to have the mgr be the only entity allowed to edit the configmap. Queued items
+could be removed once Ceph is aware a device has been added, deleted, or replaced. The mgr can be
+aware of updates to node device lists when the timestamps for all nodes' reports are after the
+timestamp for the request to list devices. In this case, Rook would have to keep internal track of
+which queue items it had already completed so that it could continue working on queue items while
+the Ceph mgr has not yet removed items from the queue.
+>
+> Another option would be to handle adding/removing queue items only by using patches to the
+configmap. I'm not certain this is possible, but it could be investigated.
+
+
+## Deprecation plan
+The first step in the deprecation plan is to have the new design in-place. Once the new design is
+available, the new design should always be used for any new OSD devices. The overall deprecation
+plan calls for pre-existing user OSDs to be migrated from the newly deprecated OSDs to OSDs using
+the new design.
+
+Regardless of progress on completing this design, users should be informed in the release notes and
```

Comment:
Again, I don't think this is possible nor reasonable to ask. Unfortunately, this is our burden/legacy and we have to deal with that. Users shouldn't suffer from bad decisions from the past :-/.

Reply:
We could still allow these old OSDs possibly. My question is whether remove operations would be different for legacy disks versus "new" disks. I suppose in the end Rook just needs to remove the k8s resources and the orchestrator will handle the OSD removal from CRUSH or the osd `out`-ing.